File: lava_reports/lava_reports/report/lava_audit_report/lava_audit_report.py
import frappe
from frappe import _
from frappe.utils import get_datetime
from frappe.query_builder.functions import DateFormat



def execute(filters=None):
    filters = filters or {}

    from_date = filters.get('from_date')
    to_date = filters.get('to_date')

    if not from_date or not to_date:
        frappe.throw(_('From Date and To Date are required'))

    from_date = get_datetime(from_date).date()
    to_date = get_datetime(to_date).date()

    if from_date > to_date:
        frappe.throw(_("From Date can't be greater than To Date"))

    columns = get_columns()
    data = get_data(filters, from_date, to_date)

    return columns, data


def get_data(filters, from_date, to_date):
    log = frappe.qb.DocType('Lava Audit Log')
    user = frappe.qb.DocType('User')

    users_filter = filters.get('users')

    query = (
        frappe.qb.from_(log)
        .select(
            log.name.as_('id'),
            DateFormat(log.date, '%a, %d %b %Y').as_('date'),
            log.user,
            user.full_name,
            log.document_type,
            log.document_name.as_('document_reference'),
            log.action
        )
        .left_join(user)
        .on(log.user == user.name)
        .where(log.date >= from_date)
        .where(log.date <= to_date)
    )

    if users_filter:
        query = query.where(log.user.isin(users_filter))

    return query.run(as_dict=True)

def get_columns():
    return [
        {
            'label': _('Audit ID'),
            'fieldname': 'id',
            'fieldtype': 'Link',
            'options': 'Lava Audit Log',
            'width': 150,
        },
        {
            'label': _('Date'),
            'fieldname': 'date',
            'fieldtype': 'Data',
            'width': 220,
        },
        {
            'label': _('User'),
            'fieldname': 'full_name',
            'fieldtype': 'Data',
            'width': 200,
        },
        {
            'label': _('Action'),
            'fieldname': 'action',
            'fieldtype': 'Data',
            'width': 120,
        },
        {
            'label': _('Document Type'),
            'fieldname': 'document_type',
            'fieldtype': 'Data',
            'width': 200,
        }
    ]

File: lava_reports/lava_reports/report/lava_audit_report/lava_audit_report.js
// Copyright (c) 2026, lavaloon and contributors
// For license information, please see license.txt

frappe.query_reports['Lava Audit Report'] = {
    'filters': [
        {
            'fieldname': 'from_date',
            'label': __('From Date'),
            'fieldtype': 'Date',
            'reqd': 1,
            'default': frappe.datetime.add_months(frappe.datetime.get_today(), -1)
        },
        {
            'fieldname': 'to_date',
            'label': __('To Date'),
            'fieldtype': 'Date',
            'reqd': 1,
            'default': frappe.datetime.get_today()
        },
        {
            'fieldname': 'users',
            'label': __('User'),
            'fieldtype': 'MultiSelectList',
			get_data: function(txt) {
				return frappe.db.get_link_options("User", txt);
			}
        }
    ],
	"formatter": function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		if (column.fieldname === "full_name" && data && data.creator) {
			value = `<a href="/app/user/${data.creator}"
				data-doctype="User"
				data-name="${data.creator}"
				data-value="${data.full_name}">${value}</a>`;
		}

		if (column.fieldname === "document_type" && data && data.document_reference) {
            const doctype_slug = data.document_type.toLowerCase().replace(/ /g, "-");
			value = `<a href="/app/${doctype_slug}/${data.document_reference}"
				data-doctype="${data.document_type}"
				data-name="${data.document_reference}"
				data-value="${data.document_reference}">${value}</a>`;
		}

		return value;
	}
};
